File: App/Resources/AlgorithmDetails/minmaxheapsort/minmaxheapsort.py
def bit_length(value):
    length = 0
    while value > 0:
        value >>= 1
        length += 1
    return length


def is_min_level(index):
    return bit_length(index + 1) % 2 == 1


def better_than(a, b, min_level):
    return a < b if min_level else a > b


def downheap(array, start, size):
    i = start
    while True:
        min_level = is_min_level(i)
        left = 2 * i + 1
        right = 2 * i + 2
        if left >= size:
            return
        winner = left
        if right < size and better_than(array[right], array[winner], min_level):
            winner = right
        base = 4 * i + 3
        for offset in range(4):
            gc = base + offset
            if gc < size and better_than(array[gc], array[winner], min_level):
                winner = gc
        is_grandchild = winner >= base
        extreme = better_than(array[winner], array[i], min_level)
        if not is_grandchild:
            if extreme:
                array[i], array[winner] = array[winner], array[i]
            return
        if extreme:
            array[i], array[winner] = array[winner], array[i]
        else:
            return
        parent = (winner - 1) // 2
        if min_level:
            if array[winner] > array[parent]:
                array[parent], array[winner] = array[winner], array[parent]
        else:
            if array[winner] < array[parent]:
                array[parent], array[winner] = array[winner], array[parent]
        i = winner


def heapify(array, length):
    for i in range((length - 1) // 2, -1, -1):
        downheap(array, i, length)


def store_max(array, heap_size):
    if heap_size <= 1:
        return heap_size
    imax = 1
    if heap_size > 2 and array[2] > array[1]:
        imax = 2
    last = heap_size - 1
    array[imax], array[last] = array[last], array[imax]
    new_size = last
    if imax < new_size:
        downheap(array, imax, new_size)
    return new_size


def sort(arr):
    n = len(arr)
    if n <= 1:
        return
    heapify(arr, n)
    heap_size = n
    for _ in range(n - 1):
        heap_size = store_max(arr, heap_size)


if __name__ == "__main__":
    array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56]
    sort(array)
    print(array)

File: App/Resources/AlgorithmDetails/minmaxheapsort/minmaxheapsort.java
import java.util.Arrays;

public final class minmaxheapsort {
  static int bitLength(int value) {
    int length = 0;
    while (value > 0) {
      value >>= 1;
      length++;
    }
    return length;
  }

  static boolean isMinLevel(int index) {
    return bitLength(index + 1) % 2 == 1;
  }

  static boolean betterThan(int a, int b, boolean minLevel) {
    return minLevel ? (a < b) : (a > b);
  }

  static void downheap(int[] arr, int start, int size) {
    int i = start;
    while (true) {
      boolean minLevel = isMinLevel(i);
      int left = 2 * i + 1;
      int right = 2 * i + 2;
      if (left >= size) {
        return;
      }
      int winner = left;
      if (right < size && betterThan(arr[right], arr[winner], minLevel)) {
        winner = right;
      }
      int base = 4 * i + 3;
      for (int offset = 0; offset < 4; offset++) {
        int gc = base + offset;
        if (gc < size && betterThan(arr[gc], arr[winner], minLevel)) {
          winner = gc;
        }
      }
      boolean isGrandchild = winner >= base;
      boolean extreme = betterThan(arr[winner], arr[i], minLevel);
      if (!isGrandchild) {
        if (extreme) {
          int temp = arr[i];
          arr[i] = arr[winner];
          arr[winner] = temp;
        }
        return;
      }
      if (extreme) {
        int temp = arr[i];
        arr[i] = arr[winner];
        arr[winner] = temp;
      } else {
        return;
      }
      int parent = (winner - 1) / 2;
      if (minLevel) {
        if (arr[winner] > arr[parent]) {
          int temp = arr[parent];
          arr[parent] = arr[winner];
          arr[winner] = temp;
        }
      } else {
        if (arr[winner] < arr[parent]) {
          int temp = arr[parent];
          arr[parent] = arr[winner];
          arr[winner] = temp;
        }
      }
      i = winner;
    }
  }

  static void heapify(int[] arr, int length) {
    for (int i = (length - 1) / 2; i >= 0; i--) {
      downheap(arr, i, length);
    }
  }

  static int storeMax(int[] arr, int heapSize) {
    if (heapSize <= 1) {
      return heapSize;
    }
    int imax = 1;
    if (heapSize > 2 && arr[2] > arr[1]) {
      imax = 2;
    }
    int last = heapSize - 1;
    int temp = arr[imax];
    arr[imax] = arr[last];
    arr[last] = temp;
    int newSize = last;
    if (imax < newSize) {
      downheap(arr, imax, newSize);
    }
    return newSize;
  }

  static void sort(int[] arr) {
    int n = arr.length;
    if (n <= 1) {
      return;
    }
    heapify(arr, n);
    int heapSize = n;
    for (int i = 0; i < n - 1; i++) {
      heapSize = storeMax(arr, heapSize);
    }
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
